examples/command_bot.rs

```diff
@@ -0,0 +1,99 @@
+use std::ops::Deref;
+use std::sync::Arc;
+use std::{env, process::exit};
+use url::Url;
+
+use matrix_sdk::{
+    self,
+    events::room::message::{MessageEvent, MessageEventContent, TextMessageEventContent},
+    AsyncClient, AsyncClientConfig, EventEmitter, Room, SyncSettings,
+};
+use tokio::sync::Mutex;
+
+struct CommandBot {
+    client: AsyncClient,
+}
+
+impl CommandBot {
+    pub fn new(client: AsyncClient) -> Self {
+        Self { client }
+    }
+}
+
+#[async_trait::async_trait]
+impl EventEmitter for CommandBot {
+    async fn on_room_message(&mut self, room: Arc<Mutex<Room>>, event: Arc<Mutex<MessageEvent>>) {
```

Comment:
I missed that completely during the review of the event emitter, but is there a reason we're putting the events into a `Mutex`?

Reply:
No, they should just be Arc right no mutation just inspection. they come out of the hashmap as an Arc/Mutex that's ok right?

Reply:
Sorry, what hash-map are you referring to here?

Reply:
oops yea the `pub joined_rooms: HashMap<RoomId, Arc<Mutex<Room>>>,` on `BaseClient`

Reply:
`room: Arc<Mutex<Room>>, event: Arc<Mutex<MessageEvent>>` works as just 
`room: &Room, event: &MessageEvent` with the `EventEmitter`

I used a `tokio::sync::RwLock` for the `Room` for the  `pub joined_rooms: HashMap<RoomId, Arc<Mutex<Room>>>` in `BaseClient` so `RwLock::read` could be used instead of `lock`ing every time.

Reply:
No, `Arc<Mutex<Room>>` won't deref to &Room, but the guard that we get after we call `lock()` or `read()` will.

With this you moved the room locking into the event emitter, this now means that there is no way for the user to avoid the deadlock if they call any method that locks the room as well.

You'll notice that the bot works as expected, that is it won't deadlock, if you don't enable the encryption feature.

We need this
```
diff --git a/examples/command_bot.rs b/examples/command_bot.rs
index c300428..a8cd5db 100644
--- a/examples/command_bot.rs
+++ b/examples/command_bot.rs
@@ -1,3 +1,4 @@
+use std::sync::Arc;
 use std::{env, process::exit};
 use url::Url;
 
@@ -6,7 +7,7 @@ use matrix_sdk::{
     events::room::message::{MessageEvent, MessageEventContent, TextMessageEventContent},
     AsyncClient, AsyncClientConfig, EventEmitter, Room, SyncSettings,
 };
-use tokio::sync::Mutex;
+use tokio::sync::{Mutex, RwLock};
 
 struct CommandBot {
     client: Mutex<AsyncClient>,
@@ -22,7 +23,7 @@ impl CommandBot {
 
 #[async_trait::async_trait]
 impl EventEmitter for CommandBot {
-    async fn on_room_message(&self, room: &Room, event: &MessageEvent) {
+    async fn on_room_message(&self, room: Arc<RwLock<Room>>, event: &MessageEvent) {
         let msg_body = if let MessageEvent {
             content: MessageEventContent::Text(TextMessageEventContent { body: msg_body, .. }),
             ..
@@ -40,7 +41,10 @@ impl EventEmitter for CommandBot {
                 formatted_body: None,
                 relates_to: None,
             });
-            let room_id = &room.room_id;
+
+            let room_id = {
+                room.read().await.room_id.clone()
+            };
 
             println!("sending");
 
diff --git a/src/base_client.rs b/src/base_client.rs
index 5c8a5bf..719e6f5 100644
--- a/src/base_client.rs
+++ b/src/base_client.rs
@@ -615,7 +615,7 @@ impl Client {
             RoomEvent::RoomMessage(msg) => {
                 if let Some(ee) = &self.event_emitter {
                     if let Some(room) = self.get_room(&room_id) {
-                        ee.on_room_message(room.read().await.deref(), &msg).await;
+                        ee.on_room_message(room.clone(), &msg).await;
                     }
                 }
             }
diff --git a/src/crypto/olm.rs b/src/crypto/olm.rs
index 1fed769..c394086 100644
--- a/src/crypto/olm.rs
+++ b/src/crypto/olm.rs
@@ -597,10 +597,13 @@ impl std::fmt::Debug for OutboundGroupSession {
 
 #[cfg(test)]
 mod test {
-    use crate::crypto::olm::Account;
+    use std::collections::HashMap;
+    use std::convert::TryFrom;
+
+    use crate::crypto::olm::{Account, OutboundGroupSession, InboundGroupSession};
+    use crate::identifiers::RoomId;
     use olm_rs::session::OlmMessage;
     use ruma_client_api::r0::keys::SignedKey;
-    use std::collections::HashMap;
 
     #[test]
     fn account_creation() {
diff --git a/src/event_emitter/mod.rs b/src/event_emitter/mod.rs
index 6ee62d1..3abb715 100644
--- a/src/event_emitter/mod.rs
+++ b/src/event_emitter/mod.rs
@@ -32,6 +32,9 @@ use crate::events::{
 };
 use crate::models::Room;
 
+use std::sync::Arc;
+use tokio::sync::RwLock;
+
 /// This trait allows any type implementing `EventEmitter` to specify event callbacks for each event.
 /// The `AsyncClient` calls each method when the corresponding event is received.
 ///
@@ -88,7 +91,7 @@ pub trait EventEmitter: Send + Sync {
     /// Fires when `AsyncClient` receives a `RoomEvent::RoomAvatar` event.
     async fn on_room_avatar(&self, _: &Room, _: &AvatarEvent) {}
     /// Fires when `AsyncClient` receives a `RoomEvent::RoomMessage` event.
-    async fn on_room_message(&self, _: &Room, _: &MessageEvent) {}
+    async fn on_room_message(&self, _: Arc<RwLock<Room>>, _: &MessageEvent) {}
     /// Fires when `AsyncClient` receives a `RoomEvent::RoomMessageFeedback` event.
     async fn on_room_message_feedback(&self, _: &Room, _: &FeedbackEvent) {}
     /// Fires when `AsyncClient` receives a `RoomEvent::RoomRedaction` event.
@@ -152,7 +155,7 @@ mod test {
         async fn on_room_avatar(&self, _: &Room, _: &AvatarEvent) {
             self.0.lock().await.push("avatar".to_string())
         }
-        async fn on_room_message(&self, _: &Room, _: &MessageEvent) {
+        async fn on_room_message(&self, _: Arc<RwLock<Room>>, _: Arc<MessageEvent>) {
             self.0.lock().await.push("message".to_string())
         }
         async fn on_room_message_feedback(&self, _: &Room, _: &FeedbackEvent) {
```

Notice how we get the room_id, we lock inside a separate scope that releases the lock immediately after the scope is left. The room_id is cloned so it isn't a reference to a now locked room struct.

The callback get a separate clone of the `Arc<Mutex<Room>>`.

Reply:
> oops yea the `pub joined_rooms: HashMap<RoomId, Arc<Mutex<Room>>>,` on `BaseClient`

The rooms come out of this hashmap, they should be `Arc<Mutex<Room>>`, the events on the other hand come out of a sync response, and are owned.
 
We clone them so we don't need to take them out of the response, but we put them behind an `Arc` so we don't need to clone the whole event for every callback.

Reply:
Do you want the rooms to be `Arc<RwLock<Room>>` as they are in the git diff or `Arc<Mutex<Room>>`?

> We clone them so we don't need to take them out of the response, but we put them behind an Arc so we don't need to clone the whole event for every callback.

is this what you want for this
```
RoomEvent::RoomMember(mem) => {
    if let Some(ee) = &self.event_emitter {
        if let Some(room) = self.get_room(&room_id) {
            ee.on_room_member(Arc::clone(&room), Arc::new(mem.clone())).await;
        }
    }
}
```
or is that ok being a &MemberEvent

Reply:
I think it doesn't matter much if they are in a `RwLock` or a `Mutex`, but let's use a `RwLock` since it's technically allows for better concurrency.

You can just clone the `Arc<Mutex<Room>>` dirrectly
```rust
ee.on_room_message(room.clone(), &msg).await;
```
The events can be borrows of the `Arc` since they will deref to that naturally, if it doesn't make any other problems (I don't think it should) it's fine to make them borrows and hide the fact that they are behind an `Arc`

Reply:
I always do it that way so its clear its an Arc clone and not just a clone. I'm fine changing it though :)

Reply:
Ah ok, I'm fine with it either way.